pkg/ruler/remotequerier.go

```diff
@@ -211,6 +222,9 @@ func (q *RemoteQuerier) query(ctx context.Context, query string, ts time.Time, l
 		}
 	}
 
+	ctx, cancel := context.WithTimeout(ctx, q.timeout)
```

Comment:
I would move this whole block `:225-:236` into a new function and then call it from `query()` and `Read()`, otherwise this portion is duplicate and this would also make it easier to reason about where the timeout starts to apply.

Potentially the initialization of the middlewares could also be in that new function, but I'm not sure if this wouldn't dilute it's scope too much.

Reply:
This was actually how I had it to start with, but it felt over complicated. I'll discuss with @ortuman if we want to do any refactoring here in a follow up PR.